File: 02-Mathematics/03-Combinatorics/cpp/unique_paths.cpp
/*
 * Problem: Unique Paths (nCr approach)
 * Source:  LeetCode 62 - https://leetcode.com/problems/unique-paths/
 * 
 * Approach: Answer is C(m+n-2, m-1). Use combinatorics.
 * Time Complexity:  O(min(m, n))
 * Space Complexity: O(1)
 * 
 * Tags: combinatorics, math
 */

#include <bits/stdc++.h>
using namespace std;

int uniquePaths(int m, int n) {
    // C(m+n-2, m-1)
    long long result = 1;
    int total = m + n - 2;
    int choose = min(m - 1, n - 1);
    
    for (int i = 0; i < choose; i++) {
        result = result * (total - i) / (i + 1);
    }
    
    return (int) result;
}

int main() {
    cout << "uniquePaths(3, 7) = " << uniquePaths(3, 7) << endl; // 28
    cout << "uniquePaths(3, 3) = " << uniquePaths(3, 3) << endl; // 6
    return 0;
}

File: 02-Mathematics/03-Combinatorics/java/UniquePaths.java
/*
 * Problem: Unique Paths
 * Source:  LeetCode 62
 * Tags: combinatorics, math
 */

public class UniquePaths {
    
    public static int uniquePaths(int m, int n) {
        long result = 1;
        int total = m + n - 2;
        int choose = Math.min(m - 1, n - 1);
        for (int i = 0; i < choose; i++) {
            result = result * (total - i) / (i + 1);
        }
        return (int) result;
    }
    
    public static void main(String[] args) {
        System.out.println("uniquePaths(3,7) = " + uniquePaths(3, 7)); // 28
        System.out.println("uniquePaths(3,3) = " + uniquePaths(3, 3)); // 6
    }
}
